crates/sui-core/src/transaction_driver/mod.rs

```diff
@@ -0,0 +1,256 @@
+    state: Mutex<State>,
+    metrics: Arc<TransactionDriverMetrics>,
+}
+
+impl<A> TransactionDriver<A>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    pub fn new(
+        authority_aggregator: Arc<AuthorityAggregator<A>>,
+        reconfig_observer: Arc<dyn ReconfigObserver<A> + Sync + Send>,
+        metrics: Arc<TransactionDriverMetrics>,
+    ) -> Arc<Self> {
+        let driver = Arc::new(Self {
+            authority_aggregator: ArcSwap::new(authority_aggregator),
+            state: Mutex::new(State::new()),
+            metrics,
+        });
+        driver.enable_reconfig(reconfig_observer);
+        driver
+    }
+
+    pub async fn submit_transaction(
+        &self,
+        // TODO(fastpath): Replace with TD request/response type once hooked up to transaction orchestrator
+        request: ExecuteTransactionRequestV3,
+        options: SubmitTransactionOptions,
+    ) -> Result<ExecuteTransactionResponseV3, TransactionDriverError> {
+        let tx_digest = request.transaction.digest();
+        let is_single_writer_tx = !request.transaction.contains_shared_object();
+        let timer = Instant::now();
+        loop {
+            match self.submit_transaction_once(&request, &options).await {
+                Ok(resp) => {
+                    let settlement_finality_latency = timer.elapsed().as_secs_f64();
+                    self.metrics
+                        .settlement_finality_latency
+                        .with_label_values(&[if is_single_writer_tx {
+                            TX_TYPE_SINGLE_WRITER_TX
+                        } else {
+                            TX_TYPE_SHARED_OBJ_TX
+                        }])
+                        .observe(settlement_finality_latency);
+                    return Ok(resp);
+                }
+                Err(e) => {
+                    tracing::warn!("Failed to submit transaction {tx_digest}: {}", e);
+                    sleep(Duration::from_secs(1)).await;
+                }
+            }
+        }
+    }
+
+    async fn submit_transaction_once(
+        &self,
+        request: &ExecuteTransactionRequestV3,
+        options: &SubmitTransactionOptions,
+    ) -> Result<ExecuteTransactionResponseV3, TransactionDriverError> {
+        // Store the epoch number; we read it from the votes and use it later to create the certificate.
+        let auth_agg = self.authority_aggregator.load();
+        let committee = auth_agg.committee.clone();
+        let epoch = committee.epoch();
+        let tx = request.transaction.clone();
+
+        // Send the transaction to a random validator.
+        let clients = auth_agg.authority_clients.iter().collect::<Vec<_>>();
+        let (name, client) = clients.choose(&mut rand::thread_rng()).unwrap();
+        let response = timeout(
+            Duration::from_secs(2),
+            client.submit_transaction(
+                SubmitTransactionRequest {
+                    transaction: bcs::to_bytes(&tx)
```

Comment:
It is a bit strange that bcs serde happens within transaction driver. Is that intentional?

Reply:
Transaction driver grpc req & resp messages use proto, so to use a different message type for the inner txn, they need to be parsed here or higher in the stack.